File: apps/etf-service/src/libs/hongsehuojian.ts
import type { SecurityConfig } from "../config/securities";
import { KLINE_COUNT, KLINE_ENDPOINT, PERIOD_DAY } from "../config/securities";
import { parseKlineJson } from "./kline-parser";
import { getShanghaiToday, toUrlDate } from "../utils/date";

/** 构造红色火箭 K 线 API 的完整 URL，包含所有查询参数 */
export const buildKlineUrl = (params: { security: SecurityConfig; now?: Date }): URL => {
  const now = params.now ?? new Date();
  // 使用上海时区当前日期作为 begin 参数（远端返回倒序数据）
  const today = getShanghaiToday(now);
  const url = new URL(KLINE_ENDPOINT);
  url.search = new URLSearchParams({
    securityCode: params.security.symbol,
    period: PERIOD_DAY,
    count: KLINE_COUNT,
    begin: toUrlDate(today),
    adjust: params.security.adjust,
    // ts 参数用于绕过 CDN 缓存
    ts: String(now.getTime()),
  }).toString();
  return url;
};

/**
 * 从红色火箭 API 抓取并解析 K 线数据。
 * 返回原始条数（非裁剪后），由上层 service 负责裁剪 T-1 之后的数据。
 */
export const fetchRemoteKlineBars = async (params: { security: SecurityConfig }) => {
  const url = buildKlineUrl({ security: params.security });
  const response = await fetch(url, {
    headers: {
      accept: "application/json",
    },
  });

  if (!response.ok) {
    throw new Error(
      `${params.security.name} 下载失败: HTTP ${response.status} ${response.statusText}`,
    );
  }

  const text = await response.text();
  return parseKlineJson({
    text,
    fileName: url.toString(),
    fallbackSymbol: params.security.symbol,
    adjType: params.security.adjType,
  });
};

File: apps/etf-service/internal/integration/hongsehuojian.go
package integration

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"time"

	"kxh-awesome/etf-service/internal/config"
	"kxh-awesome/etf-service/internal/marketdata"
)

type HongsehuojianClient struct {
	httpClient *http.Client
}

func NewHongsehuojianClient(httpClient *http.Client) *HongsehuojianClient {
	if httpClient == nil {
		httpClient = http.DefaultClient
	}
	return &HongsehuojianClient{httpClient: httpClient}
}

func (c *HongsehuojianClient) FetchRemoteKlineBars(ctx context.Context, security config.SecurityConfig) (*marketdata.ParsedKlineData, error) {
	requestURL := BuildKlineURL(security, time.Now())
	request, err := http.NewRequestWithContext(ctx, http.MethodGet, requestURL.String(), nil)
	if err != nil {
		return nil, err
	}
	request.Header.Set("accept", "application/json")

	response, err := c.httpClient.Do(request)
	if err != nil {
		return nil, err
	}
	defer response.Body.Close()

	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return nil, fmt.Errorf("%s 下载失败: HTTP %d %s", security.Name, response.StatusCode, response.Status)
	}

	body, err := io.ReadAll(response.Body)
	if err != nil {
		return nil, err
	}

	return marketdata.ParseKlineJSON(string(body), requestURL.String(), security.Symbol, security.AdjType)
}

func BuildKlineURL(security config.SecurityConfig, now time.Time) *url.URL {
	requestURL, _ := url.Parse(config.KlineEndpoint)
	today := marketdata.ShanghaiToday(now)
	query := url.Values{}
	query.Set("securityCode", security.Symbol)
	query.Set("period", config.PeriodDay)
	query.Set("count", config.KlineCount)
	query.Set("begin", marketdata.ToURLDate(today))
	query.Set("adjust", security.Adjust)
	query.Set("ts", strconv.FormatInt(now.UnixMilli(), 10))
	requestURL.RawQuery = query.Encode()
	return requestURL
}
